src/nccl_ofi_rdma.c

```diff
@@ -4245,6 +4592,71 @@ static int insert_rdma_send_req_into_msgbuff(nccl_net_ofi_rdma_send_comm_t *s_co
 	return 0;
 }
 
+static int post_rma_write_op(nccl_net_ofi_rdma_req_t *req)
+{
+	nccl_net_ofi_rdma_send_comm_t *s_comm = (nccl_net_ofi_rdma_send_comm_t *)req->comm;
+	nccl_net_ofi_rdma_send_comm_rail_t *comm_rail = get_send_comm_rail(s_comm, 0);
+	rdma_req_rma_op_data_t *rma_op_data = get_rma_op_data(req, NCCL_OFI_RDMA_WRITE);
+	struct fid_mr *rail_mr_handle = rma_op_data->buff_mr_handle->mr[0];
+	void *desc = fi_mr_desc(rail_mr_handle);
+
+	ssize_t rc;
+	/* Post RMA write */
+	rc = fi_write(comm_rail->local_ep, rma_op_data->buff,
+		      rma_op_data->buff_len, desc,
+		      comm_rail->remote_addr,
+		      rma_op_data->remote_buff,
+		      rma_op_data->remote_mr_key, req);
+
+	if ((rc != 0) && (rc != -FI_EAGAIN)) {
+		NCCL_OFI_WARN("fi_write failed; RC: %zd, Error: %s",
+			      rc, fi_strerror(-rc));
+	}
+
+	return rc;
+}
+
+static int post_rma_write_data_op(nccl_net_ofi_rdma_req_t *req)
+{
+	nccl_net_ofi_rdma_send_comm_t *s_comm = (nccl_net_ofi_rdma_send_comm_t *)req->comm;
+	nccl_net_ofi_rdma_send_comm_rail_t *comm_rail = get_send_comm_rail(s_comm, 0);
+	rdma_req_rma_op_data_t *rma_op_data = get_rma_op_data(req, NCCL_OFI_RDMA_WRITE_DATA);
+	ssize_t rc;
+
+	struct iovec iov;
+	struct fi_msg_rma msg;
+	struct fi_rma_iov rma_iov;
+
+	/* Set up the iovec */
+	iov.iov_base = rma_op_data->buff;
+	iov.iov_len = rma_op_data->buff_len;
+
+	/* Set up the rma_iov */
+	rma_iov.addr = rma_op_data->remote_buff;
+	rma_iov.len = rma_op_data->buff_len;
+	rma_iov.key = rma_op_data->remote_mr_key;
+
+	/* Initialize the message */
+	msg.msg_iov = &iov;
+	msg.desc = NULL;
```

Comment:
1. Why is `desc` NULL here, but non-NULL in `post_rma_write_op`?
2. For providers that set `FI_MR_LOCAL` (like EFA provider), it isn't valid to set `desc` to NULL here -- it must be a pointer to the descriptor from the local memory registration. From https://ofiwg.github.io/libfabric/v1.22.0/man/fi_mr.3.html
```
FI_MR_LOCAL
When the FI_MR_LOCAL mode bit is set, applications must register all data buffers that will be accessed by the local hardware and provide a valid desc parameter into applicable data transfer operations.
```
(I believe the EFA provider will automatically register any buffers for which a NULL `desc` is provided, but this isn't standards-conforming.)